presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationThroughExchange.java

```diff
@@ -0,0 +1,312 @@
+                    .map(functionRegistry::getAggregateFunctionImplementation)
+                    .allMatch(InternalAggregationFunction::isDecomposable);
+            if (!decomposable || !allowPushThrough(node.getSource())) {
+                return context.defaultRewrite(node);
+            }
+
+            Map<Symbol, Symbol> masks = node.getMasks();
+
+            Map<Symbol, FunctionCall> finalCalls = new HashMap<>();
+            Map<Symbol, FunctionCall> intermediateCalls = new HashMap<>();
+            Map<Symbol, Signature> intermediateFunctions = new HashMap<>();
+            Map<Symbol, Symbol> intermediateMask = new HashMap<>();
+            for (Map.Entry<Symbol, FunctionCall> entry : node.getAggregations().entrySet()) {
+                Signature signature = node.getFunctions().get(entry.getKey());
+
+                Symbol intermediateSymbol = generateIntermediateSymbol(signature);
+                intermediateCalls.put(intermediateSymbol, entry.getValue());
+                intermediateFunctions.put(intermediateSymbol, signature);
+                if (masks.containsKey(entry.getKey())) {
+                    intermediateMask.put(intermediateSymbol, masks.get(entry.getKey()));
+                }
+
+                // rewrite final aggregation in terms of intermediate function
+                finalCalls.put(entry.getKey(), new FunctionCall(QualifiedName.of(signature.getName()), ImmutableList.of(intermediateSymbol.toSymbolReference())));
+            }
+
+            AggregationNode partial = new AggregationNode(
+                    idAllocator.getNextId(),
+                    node.getSource(),
+                    node.getGroupBy(),
+                    intermediateCalls,
+                    intermediateFunctions,
+                    intermediateMask,
+                    node.getGroupingSets(),
+                    PARTIAL,
+                    node.getSampleWeight(),
+                    node.getConfidence(),
+                    node.getHashSymbol());
+
+            return new AggregationNode(
+                    node.getId(),
+                    context.rewrite(node.getSource(), partial),
+                    node.getGroupBy(),
+                    finalCalls,
+                    node.getFunctions(),
+                    ImmutableMap.of(),
+                    node.getGroupingSets(),
+                    FINAL,
+                    Optional.empty(),
+                    node.getConfidence(),
+                    node.getHashSymbol());
+        }
+
+        @Override
+        public PlanNode visitExchange(ExchangeNode node, RewriteContext<AggregationNode> context)
+        {
+            AggregationNode partial = context.get();
+            if (partial == null) {
+                return context.defaultRewrite(node);
+            }
+
+            List<PlanNode> newChildren = new ArrayList<>();
+            List<List<Symbol>> inputs = new ArrayList<>();
+
+            boolean reuseSymbol = node.getSources().size() == 1;
```

Comment:
let's not worry about reusing symbols. we can always run the unalias optimizer again if we want to dedup symbols